### PR TITLE
Michaelm reply pop lemmas

### PR DESCRIPTION
**Pull request by michaelmcinerney**
This is a rebased and slightly updated version of Victor's previous PR. 

I probably should have just added some new commits, rather than squashing, to make reviewing easier. Sorry. The only significant change is in the commit that says "replyPop lemmas", and there, the only significant change is the addition of the proof of `replyPop_iflive`, though there were some small changes due to the latest spec change to `replyPop`.

### CODE REVIEW

**Review by MitchellBuckley — APPROVED** [review]
Great!!

File: proof/refine/ARM/Finalise_R.thy [hunk 1]
```diff
@@ -2326,6 +2326,342 @@ lemma replyClear_invs'[wp]:
   apply wpsimp
   sorry
 
+lemma isHeadSome:
+ "isHead (Some h) \<longleftrightarrow> (\<exists>p. h = Head p)"
+  by (clarsimp simp: isHead_def split: reply_next.split)
```
Comment (mktnk3) [inline proof/refine/ARM/Finalise_R.thy:2331]: This doesn’t seem to be used, and it is pretty much the same as existing `isHead_to_head`. So, probably we should remove this.

File: proof/refine/ARM/Finalise_R.thy [hunk 2]
```diff
@@ -2326,6 +2326,342 @@ lemma replyClear_invs'[wp]:
...
+
+lemma in_list_refs_of_replies'_E:
+  assumes in_list: "(p, tp) \<in> list_refs_of_replies' s replyPtr"
+  assumes nextR: "\<And>reply. tp = ReplyNext \<Longrightarrow> replyNext reply = Some (Next p)
+                            \<Longrightarrow> replies_of' s replyPtr = Some reply \<Longrightarrow> R"
+  assumes prevR: "\<And>reply. tp = ReplyPrev \<Longrightarrow> replyPrev reply = Some p
+                            \<Longrightarrow> replies_of' s replyPtr = Some reply \<Longrightarrow> R"
+  shows "R"
```
Comment (mktnk3) [inline proof/refine/ARM/Finalise_R.thy:2339]: This doesn’t seem to be used, either. Do we want to keep it?

File: proof/refine/ARM/Finalise_R.thy [hunk 3]
```diff
@@ -2326,6 +2326,342 @@ lemma replyClear_invs'[wp]:
...
+
+lemma updateReply_ko_at'_reply_sp:
+  "\<lbrace>ko_at' reply replyPtr\<rbrace>
+   updateReply replyPtr (replyTCB_update Map.empty)
+   \<lbrace>\<lambda>_. ko_at' (reply\<lparr>replyTCB := None\<rparr>) replyPtr\<rbrace>"
+  apply (simp add: updateReply_def)
+  apply (wpsimp wp: set_reply'.set_ko_at')
+  using ko_at'_inj by blast
+
+lemma replyUnlink_ko_at'_reply_sp:
```
Comment (mktnk3) [inline proof/refine/ARM/Finalise_R.thy:2353]: Are these sp rules (this one and the above) used somewhere? If not, probably we don’t need to keep them.

**Review by mktnk3 — APPROVED** [review]
Nice. Thanks for finishing this off. Some cleanup suggestions from me, but otherwise it looks good to go!

File: proof/refine/ARM/Finalise_R.thy (hunk 1 above)
Comment (michaelmcinerney) [reply proof/refine/ARM/Finalise_R.thy:2331]: Removed

File: proof/refine/ARM/Finalise_R.thy (hunk 3 above)
Comment (michaelmcinerney) [reply proof/refine/ARM/Finalise_R.thy:2353]: Removed

File: proof/refine/ARM/Finalise_R.thy (hunk 2 above)
Comment (michaelmcinerney) [reply proof/refine/ARM/Finalise_R.thy:2339]: I think I'll add a comment saying "FIXME RT: unused?"

Comment (mktnk3) [reply proof/refine/ARM/Finalise_R.thy:2336]: Ok. I guess we can revisit when we finish Refine.

File: lib/Monad_WP/NonDetMonadVCG.thy [hunk 4]
```diff
@@ -1214,6 +1214,15 @@ lemma hoare_vcg_conj_lift:
 lemmas hoare_vcg_conj_lift_pre_fix
   = hoare_vcg_conj_lift[where P=R and P'=R for R, simplified]
 
+lemma hoare_vcg_conj_lift':
+  assumes x: "\<lbrace>P\<rbrace> f \<lbrace>Q\<rbrace>"
+  assumes y: "\<lbrace>P'\<rbrace> f \<lbrace>Q'\<rbrace>"
+  shows      "\<lbrace>P and P'\<rbrace> f \<lbrace>\<lambda>rv s. Q rv s \<and> Q' rv s\<rbrace>"
```
Comment (corlewis) [inline lib/Monad_WP/NonDetMonadVCG.thy:1220]: I don't think this lemma is used at the moment, and I think that that adding an `and` like this is fighting against the simplifier. Is anyone aware of a particular use case for it?

Comment (michaelmcinerney) [reply lib/Monad_WP/NonDetMonadVCG.thy:1220]: I will remove this as part of my forthcoming PR for doReplyTransfer_invs’, which contains some other work cleaning up some of the lemmas in this PR

Comment (mktnk3) [reply lib/Monad_WP/NonDetMonadVCG.thy:1220]: Thanks for catching this. I was thinking about it and then forgot. I am more in favour of removing this, to be honest.

**Pull request merged**